kompile: check rules with smt-lemma labels

SchmErik opened the pull request:
Rules marked with the smt-lemma label must be structured in such a way
that all terms (aside from tokens) are labeled with `smtlib` or
`smt-hook` attributes. Without these labels, it's not possible to encode
these rules for Z3.

Fixes: #2193

Commit:
kompile: check rules with smt-lemma labels

Rules marked with the smt-lemma label must be written in a way where
all terms (aside from tokens) are labeled with `smtlib` or `smt-hook`
attributes. Without these labels, it's not possible to encode
these rules for Z3.

Commit:
add test for smt-lemma checks

Commit:
Merge branch 'master' into check-smtlemma-2193

radumereuta commented on kernel/src/main/java/org/kframework/compile/checks/CheckSmtLemmas.java:
```diff
@@ -0,0 +1,49 @@
+
+            @Override
+            public void apply(KApply k) {
+                if (!k.att().toString().matches(".*smtlib.*|.*smt-hook.*")) {
```
I don't think that's how you're supposed to check if it contains those attributes.
https://github.com/kframework/k/blob/master/kore/src/main/scala/org/kframework/attributes/Att.scala#L92

dwightguth replied:
Yeah you can use `k.att().contains(Att.SMTLIB())`

SchmErik replied:
I ended up using
```
if (m.productionsFor().isDefinedAt(k.klabel()) && !m.productionsFor().get(k.klabel()).get()
         .exists(p -> p.att().contains(Att.SMT_HOOK()) || p.att().contains(Att.SMTLIB()))) {
```
instead. The `k.att().contains(Att.SMTLIB())` doesn't work because the SMTLIB is technically in k.att().att()

radumereuta commented on kernel/src/main/java/org/kframework/compile/checks/CheckSmtLemmas.java:
```diff
@@ -0,0 +1,49 @@
+        }
+        new VisitK() {
+            @Override
+            public void apply(KRewrite k) {
```
Are you sure you have to `@Override` this? The base class seems to have this code already.
https://github.com/kframework/k/blob/master/kore/src/main/scala/org/kframework/kore/transformers.scala#L95

SchmErik replied:
removed this method all together. Thanks

dwightguth commented on kernel/src/main/java/org/kframework/compile/checks/CheckSmtLemmas.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    private void check(Rule rule){
+        if (!(rule.body() instanceof KRewrite)) {
```
this seems weird. This is not going to correctly capture whether or not the rule has a rewrite in it. What are you trying to achieve here?

SchmErik replied:
this was left over from a previous change before I found out about the visitors. Before I found out about visitors, I thought I had to do a bunch of these kinds of checks but this isn't required so I removed it. Thanks for catching this

dwightguth reviewed (changes requested):
Radu's comments look appropriate to me, but I also had one more.

Commit:
Address code review feedback

Remove the override method for KRewrite. The same logic is implemented
in the super class.

Fix logic that searches productions for labels.

Commit:
remove unnecessary check

Commit:
clarify error message

Commit:
Merge branch 'master' into check-smtlemma-2193

Commit:
update test results

Commit:
Merge branch 'check-smtlemma-2193' of github.com:SchmErik/k into check-smtlemma-2193

radumereuta reviewed (commented):
lgtm, but I'll let Dwight give the final approval.

rv-jenkins committed:
Merge branch 'master' into check-smtlemma-2193
